include/vikunja/mem/iterator/ConstantIterator.hpp

```diff
@@ -0,0 +1,267 @@
+                using value_type = DataType;
+                using pointer = DataType*;
+                using reference = DataType&;
+                using iterator_category = std::random_access_iterator_tag;
+
+                /**
+                 * @brief Constructor for the ConstantIterator
+                 * @param value The value to initialize the iterator with
+                 * @param idx The index for the iterator, default 0
+                 */
+                ConstantIterator(const DataType& value, const IdxType& idx = static_cast<IdxType>(0))
+                    : v(value)
+                    , index(idx)
+                {
+                }
+
+                /**
+                 * @brief Dereference operator to receive the stored value
+                 */
+                NODISCARD ALPAKA_FN_INLINE const DataType& operator*() const
+                {
+                    return v;
+                }
+
+                /**
+                 * @brief Index operator to get stored value at some given offset from this iterator
+                 */
+                NODISCARD ALPAKA_FN_INLINE const DataType& operator[](int) const
+                {
+                    return v;
+                }
+
+#pragma region arithmeticoperators
+                /**
+                 * @brief Postfix increment operator
+                 * @note Use prefix increment operator instead if possible to avoid copies
+                 */
+                NODISCARD ALPAKA_FN_INLINE ConstantIterator operator++()
+                {
+                    ConstantIterator cpy = *this;
+                    ++index;
+                    return cpy;
+                }
+
+                /**
+                 * @brief Prefix increment operator
+                 */
+                NODISCARD ALPAKA_FN_INLINE ConstantIterator& operator++(int)
+                {
+                    ++index;
+                    return *this;
+                }
+
+                /**
+                 * @brief Postfix decrement operator
+                 * @note Use prefix decrement operator instead if possible to avoid copies
+                 */
+                NODISCARD ALPAKA_FN_INLINE ConstantIterator operator--()
+                {
+                    ConstantIterator cpy = *this;
+                    --index;
+                    return cpy;
+                }
+
+                /**
+                 * @brief Prefix decrement operator
+                 */
+                NODISCARD ALPAKA_FN_INLINE ConstantIterator operator--(int)
+                {
+                    --index;
+                    return *this;
+                }
+
+                /**
+                 * @brief Add an index to this iterator
+                 */
+                NODISCARD ALPAKA_FN_INLINE ConstantIterator operator+(const IdxType idx) const
+                {
+                    return ConstantIterator(v, index + idx);
+                }
```

Comment:
Firstly, binary arithmetic and relational operators should be implemented as hidden friends, because it avoids certain edge cases with lookup, conversions and symmetry.

```suggestion
                friend NODISCARD ALPAKA_FN_INLINE ConstantIterator operator+(const ConstantIterator& it, IdxType idx) const
                {
                    return ConstantIterator(it.v, it.index + idx);
                }
```

Secondly, `operator+` is usually implemented in terms of `operator+=`. Not always, but usually:

```suggestion
                friend NODISCARD ALPAKA_FN_INLINE ConstantIterator operator+(ConstantIterator it, IdxType idx) const
                {
                    return it += idx;
                }
```

The same applies for other arithmetic operators.